File: kernel/src/syscalls/mmap.cpp
#include "process.h"
#include "linux/mman.h"

/* uint64_t prot_to_page_flags(int prot) {
	ASSERT(!(prot & PROT_GROWSDOWN) && !(prot & PROT_GROWSUP), "prot: %d", prot);
	uint64_t page_flags = PDE64_USER;
	if (prot == PROT_NONE)
		page_flags |= PDE64_PROTNONE;
	else {
		// There's no way of having writable but not readable pages
		if (prot & PROT_WRITE)
			page_flags |= PDE64_RW;
		if (!(prot & PROT_EXEC))
			page_flags |= PDE64_NX;
	}
	return page_flags;
}
 */
/* uintptr_t Process::do_sys_mmap(UserPtr<void*> addr, size_t length, int prot,
                               int flags, int fd, size_t offset)
{
	dbgprintf("mmap(%p, %lu, 0x%x, 0x%x, %d, %p)\n", addr, length, prot,
	          flags, fd, offset);
	if (fd != -1 && !m_files.count(fd))
		return -EBADF;

	// We're not supporting multiple threads, so MAP_SHARED can be safely
	// removed
	if (flags & MAP_SHARED) {
		flags &= ~MAP_SHARED;
		//printf_once("REMOVING MAP SHARED\n");
	}

	int supported_flags = MAP_PRIVATE | MAP_ANONYMOUS | MAP_DENYWRITE | MAP_FIXED;
	ASSERT((flags & supported_flags) == flags, "unsupported flags 0x%x", flags);

	// Parse perms
	uint64_t page_flags = prot_to_page_flags(prot);
	if (!(prot & PROT_WRITE) && fd != -1)
		page_flags |= PDE64_RW; // read only file: map as writable first

	// Round length to upper page boundary and check if we have enough memory
	size_t length_upper = PAGE_CEIL(length);
	if (length_upper == 0 || !Mem::Virt::enough_free_memory(length_upper))
		return -ENOMEM;

	// Allocate memory
	void* ret;
	if (flags & MAP_FIXED) {
		if (addr.is_null() || ((addr.flat() & PTL1_MASK) != addr.flat()))
			return -EINVAL;
		ret = addr.ptr();
		Mem::Virt::alloc(ret, length_upper, page_flags);
	} else {
		ret = Mem::Virt::alloc(length_upper, page_flags);
	}

	// If a file descriptor was specified, copy its content to memory
	if (fd != -1) {
		const FileDescription& f = m_files[fd];
		// User seems to be allowed to map beyond the file limits (when
		// offset + length > f.size()). Let's see if offset > f.size() is
		// supposed to be allowed.
		ASSERT(offset <= f.size(), "offset OOB: %p / %p", offset, f.size());
		memcpy(ret, f.buf() + offset, min(f.size() - offset, length));

		// If it was read only, remove write permissions after copying content
		if (!(prot & PROT_WRITE)) {
			page_flags &= ~PDE64_RW;
			Mem::Virt::set_flags(ret, length_upper, page_flags);
		}
	}

	return (uintptr_t)ret;
} */

uint8_t prot_to_mem_perms(int prot) {
	uint8_t perms = MemPerms::None;
	if (prot & PROT_READ)
		perms |= MemPerms::Read;
	if (prot & PROT_WRITE)
		perms |= MemPerms::Write;
	if (prot & PROT_EXEC)
		perms |= MemPerms::Exec;
	return perms;
}

uintptr_t Process::do_sys_mmap(UserPtr<void*> addr, size_t length, int prot,
                               int flags, int fd, size_t offset)
{
	ASSERT(!(prot & PROT_GROWSDOWN) && !(prot & PROT_GROWSUP), "prot: %d", prot);
	dbgprintf("mmap(%p, %lu, 0x%x, 0x%x, %d, %p)\n", addr.flat(), length, prot,
	          flags, fd, offset);

	int supported_flags = MAP_PRIVATE | MAP_SHARED | MAP_ANONYMOUS |
	                      MAP_FIXED | MAP_DENYWRITE | MAP_STACK;
	ASSERT((flags & supported_flags) == flags, "unsupported flags 0x%x", flags);

	// Check given file descriptor is valid
	if (fd != -1 && !m_files.count(fd))
		return -EBADF;

	// We must return EINVAL if no length, and ENOMEM if length wraps
	if (!length)
		return -EINVAL;
	size_t length_upper = PAGE_CEIL(length);
	if (!length_upper)
		return -ENOMEM;

	bool map_private = flags & MAP_PRIVATE;
	bool map_shared = flags & MAP_SHARED;
	bool map_anonymous = flags & MAP_ANONYMOUS;
	bool map_fixed = flags & MAP_FIXED;

	// Shared and private: choose one
	if (map_shared && map_private)
		return -EINVAL;
	if (!map_shared && !map_private)
		return -EINVAL;

	// If MAP_FIXED, addr can't be null or not aligned
	if (map_fixed && (addr.is_null() || !IS_PAGE_ALIGNED(addr.flat())))
		return -EINVAL;

	// If it's a read only file, map it as writable first
	uint8_t perms = prot_to_mem_perms(prot);
	if (fd != -1)
		perms |= MemPerms::Write;

	// Note: when MAP_FIXED, a part of the specified range may already be mapped.
	uint8_t map_flags = 0;
	if (map_fixed)
		map_flags |= AddressSpace::MapFlags::DiscardAlreadyMapped;
	if (map_shared)
		map_flags |= AddressSpace::MapFlags::Shared;

	// Map range into our address space. If it fails, MAP_FIXED is not set and
	// the given address is not null, then ignore that address and try to map
	// it wherever we can. If that fails again, then it's ENOMEM.
	Range range(addr.flat() & PTL1_MASK, length_upper);
	bool success = m_space.map_range(range, perms, map_flags);
	if (!success && !map_fixed && !addr.is_null()) {
		range.set_base(0);
		success = m_space.map_range(range, perms, map_flags);
	}
	if (!success)
		return -ENOMEM;

	// Range is already mapped
	void* ret = (void*)range.base();

	// If a file descriptor was specified, copy its content to memory
	if (fd != -1) {
		const FileDescription& f = *m_files[fd];
		// User seems to be allowed to map beyond the file limits (when
		// offset + length > f.size()). Let's see if offset > f.size() is
		// supposed to be allowed.
		ASSERT(offset <= f.size(), "offset OOB: %p / %p", offset, f.size());
		memcpy(ret, f.buf() + offset, min(f.size() - offset, length));

		// If it was read only, remove write permissions after copying content
		if (!(prot & PROT_WRITE)) {
			perms &= ~MemPerms::Write;
			m_space.set_range_perms(range, perms);
		}
	}

	return (uintptr_t)ret;
}

int Process::do_sys_munmap(UserPtr<void*> addr, size_t length) {
	// Round length to upper page boundary
	length = PAGE_CEIL(length);
	if (!length || !IS_PAGE_ALIGNED(addr.flat()))
		return -EINVAL;

	Range range(addr.flat(), length);
	if (!m_space.unmap_range(range, true))
		return -EINVAL;
	return 0;
}

int Process::do_sys_mprotect(UserPtr<void*> addr, size_t length, int prot) {
	ASSERT(!(prot & PROT_GROWSDOWN) && !(prot & PROT_GROWSUP), "prot: %d", prot);
	if (!IS_PAGE_ALIGNED(addr.flat()))
		return -EINVAL;

	// Check if length wraps
	size_t length_upper = PAGE_CEIL(length);
	if (length_upper == 0 && length != 0)
		return -ENOMEM;

	Range range(addr.flat(), length_upper);
	if (!m_space.set_range_perms(range, prot_to_mem_perms(prot)))
		return -ENOMEM;
	return 0;
}

File: kernel_zig/src/process/syscalls/mmap.zig
usingnamespace @import("../common.zig");
const mem = @import("../../mem/mem.zig");
const log = std.log.scoped(.sys_mmap);

fn protToMemPerms(prot: i32) mem.Perms {
    assert(prot & linux.PROT_GROWSDOWN == 0 and prot & linux.PROT_GROWSUP == 0);
    return mem.Perms{
        .read = (prot & linux.PROT_READ) != 0,
        .write = (prot & linux.PROT_WRITE) != 0,
        .exec = (prot & linux.PROT_EXEC) != 0,
    };
}

fn sys_mmap(
    self: *Process,
    addr: usize,
    length: usize,
    prot: i32,
    flags: i32,
    fd: linux.fd_t,
    offset: usize,
) !usize {
    log.debug("mmap(0x{x}, {}, 0x{x}, 0x{x}, {}, 0x{x}\n", .{ addr, length, prot, flags, fd, offset });

    const supported_flags = linux.MAP_PRIVATE | linux.MAP_SHARED | linux.MAP_ANONYMOUS |
        linux.MAP_FIXED | linux.MAP_DENYWRITE | linux.MAP_STACK;
    assert(flags & supported_flags == flags);

    // Check given file descriptor is valid. There's a TOCTOU vuln here, but we
    // don't have multithreading so who cares.
    if (fd != -1 and !self.files.table.contains(fd))
        return error.BadFD;

    // We must return EINVAL if no length, and ENOMEM if it overflows
    const length_aligned = mem.alignPageForwardChecked(length) catch return error.OutOfMemory;
    if (length_aligned == 0)
        return error.InvalidArgument;

    const map_private = (flags & linux.MAP_PRIVATE) != 0;
    const map_shared = (flags & linux.MAP_SHARED) != 0;
    const map_anonymous = (flags & linux.MAP_ANONYMOUS) != 0;
    const map_fixed = (flags & linux.MAP_FIXED) != 0;

    // Shared and private: choose one
    if (map_shared and map_private)
        return error.InvalidArgument;
    if (!map_shared and !map_private)
        return error.InvalidArgument;

    // If MAP_FIXED, addr can't be null or not aligned
    if (map_fixed and (addr == 0 or !mem.isPageAligned(addr)))
        return error.InvalidArgument;

    // Get permisions. If we're mapping a file, map it as writable first
    // so we can write its contents.
    var perms = protToMemPerms(prot);
    if (fd != -1)
        perms.write = true;

    const map_flags = mem.AddressSpace.MapFlags{
        .discardAlreadyMapped = map_fixed,
        .shared = map_shared,
    };

    // Perform mapping
    var ret: usize = undefined;
    var retry_anywhere: bool = false;
    if (addr != 0) {
        ret = addr;
        self.space.mapRange(addr, length_aligned, perms, map_flags) catch |err| switch (err) {
            error.OutOfMemory => return error.OutOfMemory,
            error.AlreadyMapped => {
                // The given range is already mapped. As MAP_FIXED is not set,
                // addr is just a hint, so try mapping anywhere.
                retry_anywhere = true;
            },
            error.NotUserRange => {
                // The given range is invalid. If MAP_FIXED is not set, try
                // mapping anywhere. Otherwise, we must return ENOMEM.
                if (!map_fixed) {
                    retry_anywhere = true;
                } else {
                    return error.OutOfMemory;
                }
            },
        };
    }
    if (addr == 0 or retry_anywhere) {
        // If we fail here that's a true ENOMEM.
        ret = try self.space.mapRangeAnywhere(length_aligned, perms, map_flags);
    }

    // If a file descriptor was specified, copy its content to memory
    if (fd != -1) {
        const file = self.files.table.get(fd).?;
        assert(offset <= file.size()); // I don't know if this is possible TODO check it
        const copy_length = std.math.min(file.size() - offset, length);
        std.mem.copy(u8, @intToPtr([*]u8, ret)[0..copy_length], file.buf[offset .. offset + copy_length]);

        // If it was read only, remove write permissions after copying content
        if (prot & linux.PROT_WRITE == 0) {
            perms.write = false;
            self.space.setRangePerms(ret, length_aligned, perms) catch unreachable;
        }
    }

    return ret;
}

pub fn handle_sys_mmap(
    self: *Process,
    arg0: usize,
    arg1: usize,
    arg2: usize,
    arg3: usize,
    arg4: usize,
    arg5: usize,
) !usize {
    const addr = arg0;
    const length = arg1;
    const prot = std.meta.cast(i32, arg2);
    const flags = std.meta.cast(i32, arg3);
    const fd = std.meta.cast(linux.fd_t, arg4);
    const offset = arg5;
    return sys_mmap(self, addr, length, prot, flags, fd, offset);
}

fn sys_munmap(self: *Process, addr: usize, length: usize) !void {
    const length_aligned = mem.alignPageForwardChecked(length) catch return error.InvalidArgument;
    if (!mem.isPageAligned(addr))
        return error.InvalidArgument;

    // Trying to munmap a not mapped range is not an error.
    // TODO: probably wrong, as right now it stops with the first not mapped
    // area
    self.space.unmapRange(addr, length_aligned) catch |err| switch (err) {
        error.NotMapped => {},
        error.NotUserRange => return error.InvalidArgument,
    };
}

pub fn handle_sys_munmap(self: *Process, arg0: usize, arg1: usize) !usize {
    const addr = arg0;
    const length = arg1;
    try sys_munmap(self, addr, length);
    return 0;
}

fn sys_mprotect(self: *Process, addr: usize, length: usize, prot: i32) !void {
    const length_aligned = mem.alignPageForwardChecked(length) catch return error.InvalidArgument;
    if (!mem.isPageAligned(addr))
        return error.InvalidArgument;

    // mprotect returns ENOMEM if range is invalid or if any page is not mapped
    self.space.setRangePerms(addr, length_aligned, protToMemPerms(prot)) catch |err| switch (err) {
        error.NotMapped, error.NotUserRange => return error.OutOfMemory,
    };
}

pub fn handle_sys_mprotect(self: *Process, arg0: usize, arg1: usize, arg2: usize) !usize {
    const addr = arg0;
    const length = arg1;
    const prot = std.meta.cast(i32, arg2);
    try sys_mprotect(self, addr, length, prot);
    return 0;
}
